flink/src/main/java/org/apache/iceberg/flink/sink/compact/CompactFileCommitter.java

```diff
@@ -0,0 +1,171 @@
+        emit(new EndCheckpoint(
+            endCheckpointId,
+            getRuntimeContext().getIndexOfThisSubtask(),
+            getRuntimeContext().getIndexOfThisSubtask()));
+        LOG.info("Emit a EndCheckpoint {} message to downstream operator", endCheckpointId);
+      }
+
+    } else if (value instanceof CompactCommitInfo) {
+      CompactCommitInfo message = (CompactCommitInfo) value;
+      if (taskTracker == null) {
+        taskTracker = new TaskTracker(message.getNumberOfTasks());
+      }
+      LOG.info("Receive a CompactCommitInfo (checkpoint {}, total delete {} files: {}, total add {} files: {})",
+          message.getCheckpointId(),
+          message.getCurrentDataFiles().size(),
+          Joiner.on(", ").join(message.getCurrentDataFiles().stream()
+              .map(s -> s.content() + "=>" + s.fileSizeInBytes() / 1024 / 1024 + "M").collect(Collectors.toList())),
+          message.getDataFiles().size(),
+          Joiner.on(", ").join(message.getDataFiles().stream()
+              .map(s -> s.content() + "=>" + s.fileSizeInBytes() / 1024 / 1024 + "M").collect(Collectors.toList())));
+
+      this.deletedDataFiles.addAll(message.getCurrentDataFiles());
+      this.addedDataFiles.addAll(message.getDataFiles());
+
+      boolean needCommit = taskTracker.add(message.getCheckpointId(), message.getTaskId());
+      if (needCommit) {
+        LOG.info("Need to commit: checkpoint {}, total delete {} files: {}, total add {} files: {}]",
+            message.getCheckpointId(),
+            this.deletedDataFiles.size(),
+            Joiner.on(", ").join(this.deletedDataFiles.stream()
+                .map(s -> s.content() + "=>" + s.fileSizeInBytes() / 1024 / 1024 + "M").collect(Collectors.toList())),
+            this.addedDataFiles.size(),
+            Joiner.on(", ").join(this.addedDataFiles.stream()
+                .map(s -> s.content() + "=>" + s.fileSizeInBytes() / 1024 / 1024 + "M").collect(Collectors.toList())));
+        if (!this.addedDataFiles.isEmpty()) {
+          replaceDataFiles(this.deletedDataFiles, this.addedDataFiles, message.getStartingSnapshotId());
+          LOG.info("Commit compaction information successfully, checkpoint {} ", message.getCheckpointId());
+        } else {
+          LOG.warn("addedDataFiles is null, abort commit transaction");
+        }
+        this.deletedDataFiles.clear();
+        this.addedDataFiles.clear();
+      }
+    }
+  }
+
+  private void replaceDataFiles(Iterable<DataFile> currentDataFiles, Iterable<DataFile> dataFiles,
+                                long startingSnapshotId) {
+    try {
+      RewriteFiles rewriteFiles = table.newRewrite();
+      rewriteFiles.validateFromSnapshot(startingSnapshotId)
+          .rewriteFiles(Sets.newHashSet(currentDataFiles), Sets.newHashSet(dataFiles));
+      rewriteFiles.commit();
+      table.updateProperties()
+          .set(TableProperties.WRITE_COMPACT_LAST_REWRITE_MS, Long.toString(System.currentTimeMillis()))
```

Comment:
Table properties should not be updated to keep track of this kind of information. Instead, keep it in snapshot properties by calling `set` on `rewriteFiles`.